src/include/planner/planner.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "binder/bound_statement.h"
+#include "binder/statement/select_statement.h"
+#include "binder/tokens.h"
+#include "catalog/catalog.h"
+#include "catalog/column.h"
+#include "execution/expressions/abstract_expression.h"
+#include "execution/plans/abstract_plan.h"
+
+namespace bustub {
+
+/**
+ * The planner takes a bound statement, and transforms it into the BusTub plan tree.
+ * The plan tree will be taken by the execution engine to execute the statement.
+ */
+class Planner {
+ public:
+  explicit Planner(const Catalog &catalog) : catalog_(catalog) {}
+
+  // The following parts are undocumented. One `PlanXXX` functions simply corresponds to a
+  // bound thing in the binder.
+
+  void PlanQuery(const BoundStatement &statement);
+
+  auto PlanSelect(const SelectStatement &statement) -> std::unique_ptr<AbstractPlanNode>;
+
+  auto PlanTableRef(const BoundTableRef &table_ref) -> std::unique_ptr<AbstractPlanNode>;
+
+  auto PlanExpression(const BoundExpression &expr, const AbstractPlanNode &child)
+      -> std::unique_ptr<AbstractExpression>;
+
+  /** the root plan node of the plan tree */
+  std::unique_ptr<AbstractPlanNode> plan_;
+
+ private:
+  // The collection of allocated things, owned by the planner.
+  // TODO(chi): refactor ALL OF THE ARGS TO UNIQUE_PTR/SHARED_PTR instead of guessing about the ownership.
```

Comment:
I personally really liked the [ManagedPointer](https://github.com/cmu-db/noisepage/blob/master/src/include/common/managed_pointer.h) that we had in NoisePage, which is a little like observer_ptr (which afaik is still experimental).

I think someone suggested that managed pointer might be too much for students, I personally think it might be ok but it is up to you.

That said, I'm not a fan of shared_ptr for most use cases, would prefer a rule of unique ptr to denote ownership + managed or raw pointer to denote non-ownership.